File: 400/472 - Concatenated Words.java
import java.util.ArrayList;
import java.util.HashSet;

public class Solution {
    public List<String> findAllConcatenatedWordsInADict(String[] words) {
        int i, j, len;
        HashSet<String> set = new HashSet<String>();
        List<String> ret = new ArrayList<String>();
        boolean[] dp;
        
        for (String s : words)
            set.add(s);
        
        for (String s : words) {
            len = s.length();
            
            if (len == 0)
                continue;
            
            dp = new boolean[len + 1];
            dp[0] = true;
            
            for (i = 0; i < len; ++i) {
                if (! dp[i])
                    continue;
                
                for (j = i + 1; j <= len; ++j) {
                    if (dp[j])
                        continue;
                    
                    if (j - i < len && set.contains(s.substring(i, j)))
                        dp[j] = true;
                }
            }
            
            if (dp[len])
                ret.add(s);
        }
        
        return ret;
    }
}

File: csharp/472.concatenated-words.cs
/*
 * @lc app=leetcode id=472 lang=csharp
 *
 * [472] Concatenated Words
 */

using System.Collections.Generic;

public class Solution {

    private static bool FindRecurse(ISet<string> set, string word, int i, int count) {
        int n = word.Length;
        if (i >= n && count > 1) {
            return true;
        }
        for (int j = 1; j + i <= n; ++j) {
            if (set.Contains(word.Substring(i, j)) && FindRecurse(set, word, i + j, count + 1)) {
                return true;
            }
        }
        return false;
    }

    public IList<string> FindAllConcatenatedWordsInADict(string[] words) {
        int m = words.Length;
        IList<string> ret = new List<string>();
        ISet<string> set = new HashSet<string>(m);
        SortedSet<int> lengths = new SortedSet<int>();
        foreach (string word in words) {
            int len = word.Length;
            if (len == 0) {
                continue;
            }
            set.Add(word);
            lengths.Add(len);
        }
        foreach (string word in words) {
            if (word.Length < 2 * lengths.Min) {
                continue;
            }
            if (FindRecurse(set, word, 0, 0)) {
                ret.Add(word);
            }
        }
        return ret;
    }
}

